metals/src/main/scala/scala/meta/internal/metals/codeactions/UseNamedArguments.scala

```diff
@@ -0,0 +1,164 @@
+      case Term.Select(_, x) => Some(x)
+      case Term.Apply(x, _) => findSymbolTree(x)
+      case Term.ApplyType(x, _) => findSymbolTree(x)
+      case Type.Apply(x, _) => findSymbolTree(x)
+      case Init(x, _, _) => findSymbolTree(x)
+      case _ => None
+    }
+
+    def resolveSymbol(
+        textDocumentId: l.TextDocumentIdentifier,
+        path: AbsolutePath,
+        textDocument: TextDocument,
+        symbolTreePos: Position
+    ): ResolvedSymbolOccurrence = {
+      val tdpp = new l.TextDocumentPositionParams(
+        textDocumentId,
+        symbolTreePos.toLSP.getStart
+      )
+      definitionProvider.positionOccurrence(path, tdpp, textDocument)
+    }
+
+    val methodSymbolPattern = """.*\((\+\d+)?\)\.$"""
+
+    def tweakSymbol(symbol: String): String = {
+      if (symbol.endsWith(".") && !symbol.matches(methodSymbolPattern)) {
+        /*
+         * We've probably been given a companion object symbol
+         * e.g. in the case (@@ = cursor)
+         *
+         * case class Foo(a: Int, b: Int)
+         * val x = Fo@@o(1, 2)
+         *
+         * the definition provider gives us the symbol of the companion
+         * object. This is no good to us, so we try to replace it with
+         * the case class's symbol.
+         * In other words we turn "example/Foo." into "example/Foo#"
+         */
+        symbol.stripSuffix(".") ++ "#"
+      } else {
+        symbol
+      }
+    }
+
+    def buildEdits(
+        tree: Tree,
+        paramNames: List[String]
+    ): List[l.TextEdit] = {
+      val args = tree match {
+        case Term.Apply(_, xs) => xs
+        case Init(_, _, xss) => xss.flatten
+        case _ => Nil
+      }
+      args.zip(paramNames).flatMap {
+        case (Term.Assign(_, _), _) =>
+          // already a named argument, no edit needed
+          None
+        case (term, paramName) =>
+          val position = term.pos.toLSP.getStart
+          val text = s"$paramName = "
+          val edit = new l.TextEdit(new l.Range(position, position), text)
+          Some(edit)
+      }
+    }
+
+    val path = params.getTextDocument().getUri().toAbsolutePath
+
+    Future {
+      (for {
+        rootTree <- trees.get(path)
+        methodApplyTree <- findMethodApplyOrCtorTreeUnderCursor(
+          rootTree,
+          params.getRange
+        )
+        textDocument <- semanticdbs.textDocument(path).documentIncludingStale
+        symbolTree <- findSymbolTree(methodApplyTree)
+        resolvedSymbol = resolveSymbol(
+          params.getTextDocument,
+          path,
+          textDocument,
+          symbolTree.pos
+        )
+        symbolOccurrence <- resolvedSymbol.occurrence
+        symbol = tweakSymbol(symbolOccurrence.symbol)
+        symbolDocumentation <- symbolSearch
+          .documentation(symbol)
+          .asScala
+        parameterNames = symbolDocumentation
+          .parameters()
+          .asScala
+          .map(_.displayName())
+          .toList
```

Comment:
This part here is using `symbolSearch` and `symbolDocumentation` to lookup the signature information of the method under the cursor.

This is very similar to the job `SignatureHelpProvider` does, but the main difference is that `SignatureHelpProvider` uses the presentation compiler, so it should work even with code you just typed or when the buffer doesn't parse or compile.

I also think `SignatureHelpProvider` does a very similar job at finding the enclosing `apply` (see `EnclosingApply`).

Bottom line, I see a chance of building the core of this action on top of the presentation compiler, possibly re-using some of the infrastructure provided by `SignatureHelpProvider`, or even just using that provider directly.

As a quick POC, you can maybe try invoking `compilers.signatureHelp` and see whether that's enough information to build this action.

Reply:
Paging @olafurpg, as soon as you have time

Reply:
Thanks, I'll give it a try

Reply:
@gabro Wow, that certainly simplified things!